third_party/txt/src/txt/paragraph_txt.cc

```diff
@@ -1186,79 +1197,98 @@ void ParagraphTxt::UpdateLineMetrics(const SkFontMetrics& metrics,
+    // descent is `style.height * style.font_size`.
     //
-    // Disable first line ascent modifications.
-    if (line_number == 0 && paragraph_style_.text_height_behavior &
-                                TextHeightBehavior::kDisableFirstAscent) {
-      ascent = -metrics.fAscent;
-    }
-    // Disable last line descent modifications.
-    if (line_number == line_limit - 1 &&
-        paragraph_style_.text_height_behavior &
-            TextHeightBehavior::kDisableLastDescent) {
-      descent = metrics.fDescent;
-    }
+    // The raw metrics do not add up to fontSize. The state of font
+    // metrics is a mess:
+    //
+    // Each font has 4 sets of vertical metrics:
+    //
+    // * hhea: hheaAscender, hheaDescender, hheaLineGap.
+    //     Used by Apple.
+    // * OS/2 typo: typoAscender, typoDescender, typoLineGap.
+    //     Used sometimes by Windows for layout.
+    // * OS/2 win: winAscent, winDescent.
+    //     Also used by Windows, generally will be cut if extends past
+    //     these metrics.
+    // * EM Square: ascent, descent
+    //     Not actively used, but this defines the 'scale' of the
+    //     units used.
+    //
+    // `Use Typo Metrics` is a boolean that, when enabled, prefers
+    // typo metrics over win metrics. Default is off. Enabled by most
+    // modern fonts.
+    //
+    // In addition to these different sets of metrics, there are also
+    // multiple strategies for using these metrics:
+    //
+    // * Adobe: Set hhea values to typo equivalents.
+    // * Microsoft: Set hhea values to win equivalents.
+    // * Web: Use hhea values for text, regardless of `Use Typo Metrics`
+    //     The hheaLineGap is distributed half across the top and half
+    //     across the bottom of the line.
+    //   Exceptions:
+    //     Windows: All browsers respect `Use Typo Metrics`
+    //     Firefox respects `Use Typo Metrics`.
+    //
+    // This pertains to this code in that it is ambiguous which set of
+    // metrics we are actually using via SkFontMetrics. This in turn
+    // means that if we use the raw metrics, we will see differences
+    // between platforms as well as unpredictable line heights.
+    //
+    // A more thorough explanation is available at
+    // https://glyphsapp.com/tutorials/vertical-metrics
+    //
+    // Doing this ascent/descent normalization to the EM Square allows
+    // a sane, consistent, and reasonable "blob_height" to be specified,
+    // though it breaks with what is done by any of the platforms above.
+    const bool shouldNormalizeFont =
+        style.has_height_override &&
+        !(text_height_behavior & TextHeightBehavior::kHalfLeading);
+    const double font_height =
+        shouldNormalizeFont ? style.font_size : metrics_font_height;
+
+    const double leading =
+        text_height_behavior & TextHeightBehavior::kHalfLeading
+            ? blob_height - font_height
+        : style.has_height_override ? 0.0
```

Comment:
I can't quite figure out this part of the ternary. This `0.0` will be hit when: "if not half leading, and style has a height override, then leading == 0"? Is that because we don't consider there to be any "leading" with AD scaling, instead of adding leading we just stretch the ascent and descent?

Reply:
Side note and totally personal preference so feel free to ignore: it takes me a few extra seconds to parse the nested ternary, I personally would find something like this easier to read:

```cpp
if (halfLeading) {
  leading = blob_height - font_height;
else {
  leading = style.has_height_override ? 0.0 : metrics.fLeading;
}